disorder-core/src/Disorder/Core/Gen.hs

```diff
@@ -123,3 +127,50 @@ genDeterministic' :: Int -> GenSeed -> Gen a -> a
 genDeterministic' size (GenSeed seed) (MkGen g) =
   let r = mkQCGen seed in
   g r size
+
+-- | Generate a list of a given length containing no duplicates.
+--
+-- As the provided length can be greater than the cardinality of the type
+-- (number of distinct representable values), this function is not guaranteed
+-- to terminate successfully; it will give up after 10000 consecutive failures
+-- to generate a new value.
+--
+-- `vectorOfUnique n` runs the generator with a size equal to the maximum of
+-- `n` and the current size parameter; setting this above zero is necessary to
+-- use this with sized generators from the QuickCheck test runner. The optimal
+-- value depends on the cardinality of the type, how the generator in question
+-- uses the size parameter, and the desired number of unique values.
+vectorOfUnique' :: Eq a => Int -> Int -> Gen a -> Gen [a]
+vectorOfUnique' s k g =
+  scaled (max s) $ go k [] mana
+  where
+    go 0 xs _ = pure xs
+    go _ xs 0 =
+      fail $ "vectorOfUnique' out of mana - are you trying to generate a list larger than the cardinality of the type? (" <> show (length xs) <> "/" <> show k <> ")"
+    go n xs mana' =
+      g >>= \y -> case elem y xs of
```

Comment:
I'm not sure whether it's worth doing, but I often need a list of values that are unique by a subset of `a`. Is it worth passing in `a -> a -> Bool` instead of `Eq` here? I can't promise I would use it so don't do it on my behalf.